src/util/auth/hs_auth.ts

```diff
@@ -94,58 +86,42 @@ export class RequestAuthentication {
-		let user: User;
-		try {
-			user = await this.authenticate(req, res);
-		} catch (err) {
-			// Either user was not authenticated, or an error occured during authentication
-			// In both cases we redirect them to the login
-			const queryParam: string = querystring.encode({
-				returnto: `${getConfig().hs.applicationUrl}${req.originalUrl}`
-			});
-			res.redirect(`${getConfig().hs.authUrl}/login?${queryParam}`);
-			return;
-		}
-		res.locals.authLevel = user.authLevel;
-		return next();
-	}
+	public withAuthMiddleware(router: RouterInterface, operationHandler: ExpressOpHandlerFunction): AuthMiddlewareFunction<unknown> {
+		return async (req: Request, res: Response, next: NextFunction): Promise<unknown> => {
+			const userAuth = this.authenticate(req, res);
 
-	public checkAuthLevel(req: Request, res: Response, user: User|undefined, requiredAuth: AuthLevel): boolean {
-		if (!user || user.authLevel < requiredAuth) {
-			const queryParam: string = querystring.encode({ returnto: `${getConfig().hs.applicationUrl}${req.originalUrl}` });
-			res.redirect(`${getConfig().hs.authUrl}/login?${queryParam}`);
-			return false;
-		}
-		return true;
+			const requestedUri = this.getUriFromRequest(router, operationHandler, req);
+			const resourceAuth = this.authApi.getAuthorizedResources(this.getAuthToken(req), [requestedUri]);
+
+			try {
+				const [user, permissions] = await Promise.all([userAuth, resourceAuth]);
+				if (permissions.length === 0) {
+					return this.handleUnauthorized(req, res);
+				}
+				req.user = user;
+			} catch (err) {
+				return this.handleUnauthorized(req, res);
+			}
+
+			return operationHandler(req, res, next);
+		};
 	}
 
-	public checkIsAttendee(req: Request, res: Response, next: NextFunction): void {
-		if (this.checkAuthLevel(req, res, req.user as User, AuthLevel.Attendee)) {
-			res.locals.isAttendee = true;
-			return next();
-		}
+	public handleUnauthorized(req: Request, res: Response): void {
+		const queryParam: string = querystring.encode({ returnto: `${getConfig().hs.applicationUrl}${req.originalUrl}` });
+		res.redirect(`${getConfig().hs.authUrl}/login?${queryParam}`);
 	}
 
-	public checkIsVolunteer(req: Request, res: Response, next: NextFunction): void {
-		if (this.checkAuthLevel(req, res, req.user as User, AuthLevel.Volunteer)) {
-			res.locals.isVolunteer = true;
-			return next();
-		}
+	public getAuthToken(req: Request): string {
+		return req.cookies[this.cookieName];
 	}
 
-	public checkIsOrganiser(req: Request, res: Response, next: NextFunction): void {
-		if (this.checkAuthLevel(req, res, req.user as User, AuthLevel.Organiser)) {
-			res.locals.isOrganiser = true;
-			res.locals.isVolunteer = true;
-			return next();
-		}
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	private getUriFromRequest(router: RouterInterface, operationHandler: ExpressOpHandlerFunction, _req: Request): string {
```

Comment:
We don't take into account the arguments in the request here right? 

If we don't need this kind of functionality right now, let's leave it as is (though we should create a TODO).

Reply:
Yeah, we may need the args here in the future, I'll add a TODO

I don't think we'll need it initially though